File: services/rust/poem/src/api/handlers/delete.rs
use crate::{
    api::errors::{internal_server_error, task_not_found},
    state::State as ServerState,
};
use poem::{IntoResponse, Response, http::StatusCode, web::Data};
use std::sync::{Arc, Mutex};
use tracing::{error, info, instrument, warn};
use uuid::Uuid;

#[poem::handler]
#[instrument(skip_all)]
pub async fn delete_handler(
    Data(state): Data<&Arc<Mutex<ServerState>>>,
    Data(id): Data<&Uuid>,
) -> Response {
    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.remove(&id) {
            drop(state);
            info!(
                %id,
                secret = task.secret,
                operation = task.operation.to_string(),
                method = "DELETE",
                "Removed task"
            );
            return StatusCode::NO_CONTENT.into_response();
        } else {
            drop(state);
            warn!(%id, method = "DELETE", "Task not found");
            return task_not_found();
        }
    }

    error!(%id, method = "DELETE", "Poisoned lock");

    internal_server_error()
}

File: services/typescript/hono/src/routes/handlers/delete.ts
import type { Context } from "hono";
import { z } from "zod";
import type { State } from "../../state.js";
import { getLogger } from "../../logger.js";
import { AppErrors, sendError } from "../errors.js";

export function deleteHandler(state: State) {
  return (c: Context) => {
    const logger = getLogger();

    const id = z.uuidv4().safeParse(c.req.param("id"));
    if (!id.success) {
      logger.warn({ method: c.req.method, path: c.req.path }, "Invalid path");
      return sendError(c, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);

    if (task === undefined) {
      logger.warn(
        { id: id.data, method: "DELETE", path: c.req.path },
        "Task not found"
      );
      return sendError(c, AppErrors.TaskNotFound);
    }

    state.tasks.delete(id.data);

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "DELETE"
      },
      "Removed task"
    );

    return c.body(null, 204);
  };
}
